### ISSUE TITLE
HDS fragment timeouts result in incomplete VOD

### ISSUE THREAD

**Issue by tyaremco** [chrippa/livestreamer#1215]
I use livestreamer to record HDS on demand videos.  Occasionally livestreamer will fail to retrieve a fragment due to timeout, typically the first fragments.  

`[download][..ion.WEBRIP.720p-003.flv] Written 82.2 MB (2m30s @ 591.0 KB/s)` 
`[stream.hds][error] Failed to parse fragment 1-82: Failed to read data: timed out`
`[download][..ion.WEBRIP.720p-003.flv] Written 140.9 MB (4m14s @ 801.5 KB/s)`

This happens randomly; downloading the same file again might result in a complete download or a different missing fragment.

The result is that the output file will be missing data.  I know this because the video will skip on playback and multiple download attempts often result in different file sizes.  

AFAIK there is no way to request the fragment again or even to restart the download when this error occurs.  **It would be nice if we had an option to ensure that all fragments are downloaded.**  Maybe this is not practical for live streams but a VOD should be able to be saved in their entirety, even if a fragment fails on the first attempt.

I can't tell if this is what `--hds-segment-attempts` is intended to do (I set this to 10 and still got the same behavior).

**Comment by scottbernstein-zz** [comment]
Yes, you hit the nail on the head.  --hds-segment-attempts should do exactly what you want.  Why not set it to 100 or 500 tries?  That is essentially ensuring that all segments will be downloaded if you set it high enough.

Other command line options that will likely help you here:
--hds-segment-threads (i'd set this as high as possible -- 10)
--hds-segment-timeout (well, your error specifically says that it timed out...so I'd definitely raise this as well to double/triple/quadruple, until you find something that works)

**Comment by tyaremco** [comment]
Just to confirm, when I see this:
`[stream.hds][error] Failed to parse fragment 1-82: Failed to read data: timed out`
has livestreamer already made all the attempts specified by `--hds-segment-attempts` and given up?

Can you please explain how increasing the thread count might help to avoid timeouts?

**Comment by scottbernstein-zz** [comment]
Yes, that is exactly what that error message indicates.

More threads allow more requests to be made in parallel.  Gives the program the opportunity for retries for each segment while the other segments continue.

**Comment by tyaremco** [comment]
Thank you.

After using these suggestions, I now occasionally get errors like this:
`[stream.hds][error] Failed to parse fragment 1-2: Failed to read data: [Errno 10054] An existing connection was forcibly closed by the remote host`

It's not clear whether `--hds-segment-attempts 500` covers this type of error.  Did this happen on the 500th attempt or did it happen once and the fragment was immediately dropped?
